fearless/Modules/Staking/Analytics/Common/AnalyticsViewModelFactoryBase.swift

```diff
@@ -0,0 +1,205 @@
+                daysLeftText: .init(string: subtitle),
+                tokenAmountText: "+\(tokenAmountText)",
+                usdAmountText: txTimeText
+            )
+        }
+    }
+
+    /// Overrinde in subclasses
+    func getHistoryItemTitle(data _: T, locale _: Locale) -> String {
+        ""
+    }
+
+    /// Overrinde in subclasses
+    private func getTokenAmountText(data: T, locale: Locale) -> String {
+        guard
+            let tokenDecimal = Decimal.fromSubstrateAmount(
+                data.amount,
+                precision: chain.addressType.precision
+            )
+        else { return "" }
+
+        let tokenAmountText = balanceViewModelFactory
+            .amountFromValue(tokenDecimal)
+            .value(for: locale)
+        return tokenAmountText
+    }
+
+    private func createSections(
+        rewardsData: [T],
+        locale: Locale
+    ) -> [AnalyticsRewardSection] {
+        let dateTitleFormatter = DateFormatter()
+        dateTitleFormatter.locale = locale
+        dateTitleFormatter.dateFormat = "MMM d"
+
+        let groupedByDay = rewardsData
+            .groupedBy(dateComponents: [.year, .month, .day])
+        let sortedByDay: [(Date, [T])] = groupedByDay.keys
+            .map { (key: Date) in
+                (key, groupedByDay[key]!)
+            }
+            .sorted(by: { $0.0 > $1.0 })
+
+        return sortedByDay
+            .map { date, rewards in
+                let items = createViewModelItems(rewardsData: rewards, locale: locale)
+                let title = dateTitleFormatter.string(from: date).uppercased()
+
+                return AnalyticsRewardSection(
+                    title: title,
+                    items: items
+                )
+            }
+    }
+
+    private func groupedData<T: AnalyticsViewModelItem>(
+        _ data: [T],
+        by period: AnalyticsPeriod,
+        periodDelta: Int
+    ) -> [Decimal] {
+        data.reduce(into: [Decimal](repeating: 0.0, count: period.chartBarsCount)) { array, value in
+            guard let decimal = Decimal.fromSubstrateAmount(
+                value.amount,
+                precision: chain.addressType.precision
+            ) else { return }
+
+            let timestampInterval = period.timestampInterval(periodDelta: periodDelta)
+            let distance = timestampInterval.1 - timestampInterval.0
+            let index = Int(
+                Double(value.timestamp - timestampInterval.0) / Double(distance) * Double(period.chartBarsCount)
+            )
+            array[index] += decimal
+        }
+    }
+}
+
+protocol Dated {
+    var date: Date { get }
+}
+
+private extension Array where Element: Dated {
+    func groupedBy(dateComponents: Set<Calendar.Component>) -> [Date: [Element]] {
+        let initial: [Date: [Element]] = [:]
+        let groupedByDateComponents = reduce(into: initial) { acc, cur in
+            let components = Calendar.current.dateComponents(dateComponents, from: cur.date)
```

Comment:
I suggest to path a calendar as parameter to make it consistent across whole analytics logic